### ISSUE TITLE
How should the keyId be verified?

### ISSUE THREAD

**Issue by edent** [swicg/activitypub-http-signature#30]
Suppose a malicious server sent out a message claiming to be from user Alice, but with a signature  from Bob

`signature: keyId="https://example.com/users/Bob#main-key",algorithm="rsa-sha256" ...`

And in the body have *different* actor - `"actor": "https://example.com/users/Alice"` - or `"https://example.ORG/users/Alice"`

Verifying the signature is generated by the public key from keyId doesn't verify whether keyId is the *actor's* key.

So what is the order in which keys should be retrieved and checked?

a) Get the public key from the keyId, verify the signature, get the actor's key from their webfinger, compare the keys.

b) Get the public key from the keyId, verify the signature, ignore the actor's key.

c) Get the actor's public key from their webfinger, verify the signature, get the key from the keyId, compare the keys.

d) Get the actor's public key from their webfinger, verify the signature, ignore the key from the keyId.

e) Get both keys, compare them, and only verify the signature if they match.

(Feel free to tell me I've funadmentally misunderstood something. Or that I'm overthinking. My brain is melting.)

**Comment by snarfed** [comment]
Yes! Authentication vs authorization, both underspecified in AP. 😁

My understanding is, you always want to use the HTTP Sig's `keyId` key to verify the signature. That's the authentication part, and it generally happens before authorization. Alternatively, if there's an [LD Sig](https://www.w3.org/wiki/SocialCG/ActivityPub/Authentication_Authorization#Data_integrity_proofs_/_Linked_Data_Signatures) in the body - rare, but some servers like Mastodon send them during eg [inbox forwarding](https://www.w3.org/TR/activitypub/#inbox-forwarding) - you can verify that instead of the HTTP Sig. (Details out of scope for this repo.)

Once that sig verification passes, _then_ you can move on to authorization, ie is the actor who signed the request/activity actually authorized to perform the activity, based on the _activity's_ actor, etc. Details on that here: https://www.w3.org/wiki/SocialCG/ActivityPub/Authentication_Authorization#Authorization

**Comment by snarfed** [comment]
OK, [section 2.3 in the current draft](https://swicg.github.io/activitypub-http-signature/#how-to-obtain-a-signature-s-public-key) addresses finding a given `keyId`'s public key.  Beyond that, [authorization/access control is a non-goal of the report](https://swicg.github.io/activitypub-http-signature/#non-goals), but I'm happy to hear proposals for specific language or changes if you want!

**Comment by steve-bate** [comment]
> OK, [section 2.3 in the current draft](https://swicg.github.io/activitypub-http-signature/#how-to-obtain-a-signature-s-public-key) addresses finding a given `keyId`'s public key. Beyond that, [authorization/access control is a non-goal of the report](https://swicg.github.io/activitypub-http-signature/#non-goals), but I'm happy to hear proposals for specific language or changes if you want!

Section 4 "So, what's the verdict?", mentions several Fediverse use cases for using HTTP Signatures for authorization (e.g., "Block" enforcement) even though it's explicitly a "non goal". This is potentially confusing.

It also states "[Fediverse servers] use HTTP Signatures to identify the remote user, then check if they're blocked. This works, more or less.". It's definitely "less". With authorized fetch and instance actors, this can only be done at the server/service/instance level, not the user level. That's one of the disadvantages of the technique.

**Comment by snarfed** [comment]
> Section 4 "So, what's the verdict?", mentions several Fediverse use cases for using HTTP Signatures for authorization (e.g., "Block" enforcement) even though it's explicitly a "non goal". This is potentially confusing.

True! Authorization is the primary use case for authentication in general, so it's hard to omit it from the report entirely. The point of the non-goal is mainly that we're not trying to comprehensively survey authorization as well as authentication. Good nudge though, I'll expand that language!

> It also states "[Fediverse servers] use HTTP Signatures to identify the remote user, then check if they're blocked. This works, more or less.". It's definitely "less". With authorized fetch and instance actors, this can only be done at the server/service/instance level, not the user level. That's one of the disadvantages of the technique.

Hmm! My understanding is that at least some fediverse servers enforce user-level as well as server-level blocks with authorized fetch. Eg if Alice's server enables authorized fetch, and local user Alice blocks remote user Bob, and it gets a request for a post of Alice's signed by Bob, it will reject that request.

**Comment by steve-bate** [comment]
> Hmm! My understanding is that at least some fediverse servers enforce user-level as well as server-level blocks with authorized fetch. Eg if Alice's server enables authorized fetch, and local user Alice blocks remote user Bob, and it gets a request for a post of Alice's signed by Bob, it will reject that request.

But in an authorized fetch scenario at least one (and typically both) actor fetches are performed by the instance actor so that bypasses user-level blocking (and potentially even domain-level blocking). For other fetches (like the posts you mention), I don’t know if servers like Mastodon use instance actor signing or not. My understanding is that they do not since those fetches are not typically user-specific. It seems to me that the user/actor-level signing and related authorization is more relevant to “authorized post” than to fetch.

But I agree. *If* Bob signs the fetch, it’s possible that Alice could block it. But, why would Bob sign it in that case instead of delegating the fetch signing to some unblocked actor (like the instance actor)?

**Comment by snarfed** [comment]
> But in an authorized fetch scenario at least one (and typically both) actor fetches are performed by the instance actor so that bypasses user-level blocking (and potentially even domain-level blocking)...It seems to me that the user/actor-level signing and related authorization is more relevant to “authorized post” than to fetch.

True! Good point. At least for servers that have instance actors and use them to fetch objects to avoid signature verification deadlock, you're absolutely right.

> But I agree. _If_ Bob signs the fetch, it’s possible that Alice could block it. But, why would Bob sign it in that case instead of delegating the fetch signing to some unblocked actor (like the instance actor)?

It's pretty much always the server choosing how to generate the signature, not the actor themselves, so it just depends on what the server supports and how it's configured. Regardless, good point!

At a high level, I'm very much with you that HTTP Sigs are nowhere near good enough for what the fediverse currently uses them for. I tried to be fair and reasonable in the report conclusion, but hopefully that sentiment still comes across.

**Comment by snarfed** [comment]
Tentatively closing. Feel free to re-open!
